File: HW2/Codes/22002717_Iynem_Arda.lua
refEnvChecker = "Deep Binding"

function main()
  local operationNumber = 1
  local list = {}

  -- Operation 1: Are formal and actial parameters type checked?
  print("\nOperation " .. operationNumber)
  typeChecking(1, 2)
  typeChecking(1, "2.0") -- NO TYPE CHECKING
  operationNumber = operationNumber + 1

  -- Operation 2: Are keyword (named) parameters supported?
  -- NO
  print("\nOperation " .. operationNumber)
  print("NOT SUPPORTED")
  operationNumber = operationNumber + 1

  -- Operation 3: Are default paremeters supported?
  -- NO
  print("\nOperation " .. operationNumber)
  print("NOT SUPPORTED")
  operationNumber = operationNumber + 1

  -- Operation 4: What are the paremeter passing methods provided?
  print("\nOperation " .. operationNumber)
  local tmpInt = 300

  print("\n-- Before function call --\ntmpInt in main: " ..
    tmpInt .. "\nlist in main: " ..
    table.concat(list, " "))

  parameterPassByValue(list, tmpInt)

  print("\n-- After function call --\ntmpInt in main: " ..
    tmpInt .. "\nlist in main: " ..
    table.concat(list, " "))

  operationNumber = operationNumber + 1

  -- Operation 5: Can subprograms be passed as parameters? If so, how is the referencing environment of the passed subprogram bound?
  print("\nOperation " .. operationNumber)
  local refEnvChecker = "Ad Hoc Binding"
  functionVariable = passedFunction
  subprogramPass(functionVariable, operationNumber)
  operationNumber = operationNumber + 1

end

function typeChecking(intVariable, intVariable2)
  print(intVariable + intVariable2)
end

-- all types are passed by value, but function, table, userdata and thread are reference (coroutine) types and are (imitating) passed by reference (shallow copied passed by value).
function parameterPassByValue(list, intVariable)
  intVariable = 15
  table.insert(list, "addedElement")
  list = {"MutableType_PassByReferenceChecker"}

  print("\n-- In pass-by-value function call --\ntmpInt in main: " ..
    intVariable .. "\nlist in main: " ..
    table.concat(list, " "))
end

function subprogramPass(functionVariable, intVariable)
  local refEnvChecker = "Shallow Binding"
  functionVariable(intVariable)
end

function passedFunction(intVariable)
  print("Passed function with a parameter: " .. intVariable)
  print("Binding refEnvChecker: " .. refEnvChecker)
end

main()

File: HW2/Codes/22002717_Iynem_Arda.rs
static refEnvChecker: &'static str = "Deep Binding";

fn main() {
    let mut operation_number: i32 = 1;
    let mut list: Vec<String> = vec![];

    // Operation 1: Are formal and actual parameters type checked?
    println!("\nOperation {}", operation_number);
    type_checking(1, 2);
    // type_checking(1, "2.0"); SYNTAX ERROR AT LINE 10
    operation_number += 1;

    // Operation 2: Are keyword (named) parameters supported?
    // No
    println!("\nOperation {}", operation_number);
    println!("Not Supported");
    operation_number += 1;

    // Operation 3: Are default parameters supported?
    // No
    println!("\nOperation {}", operation_number);
    println!("Not Supported");
    operation_number += 1;

    // Operation 4: What are the parameter passing methods provided?
    println!("\nOperation {} (Pass by Move)", operation_number);

    let mut tmp_int = 300;

    println!(
        "\n-- Before function call --\ntmp_int in main: {}\nlist in main: {}",
        tmp_int,
        list.join(" ")
    );

    // parameter_pass_by_value(list, tmp_int); passes the ownership which is moved to function back to the list variable to make it accessible again
    list = parameter_pass_move(list, tmp_int);

    println!(
        "\n-- After function call --\ntmp_int in main: {}\nlist in main: {}",
        tmp_int,
        list.join(" ")
    );

    println!("\nOperation {} (Pass by Borrow)", operation_number);

    println!(
        "\n-- Before function call --\ntmp_int in main: {}\nlist in main: {}",
        tmp_int,
        list.join(" ")
    );

    parameter_pass_borrow(&mut list, &mut tmp_int);

    println!(
        "\n-- After function call --\ntmp_int in main: {}\nlist in main: {}",
        tmp_int,
        list.join(" ")
    );

    operation_number += 1;

    // Operation 5: Can subprograms be passed as parameters? If so, how is the referencing environment of the passed subprogram bound?
    println!("\nOperation {}", operation_number);
    // let refEnvChecker: &str = "Ad Hoc Binding"; SYNTAX ERROR WHEN GLOBAL REFENVCHECKER IS COMMENTED OUT
    subprogramPass(passedFunction, operation_number);
    operation_number += 1;
}

fn type_checking(int_variable: i32, int_variable2: i32) {
    println!("{}", int_variable + int_variable2);
}

fn parameter_pass_move(mut list: Vec<String>, mut int_variable: i32) -> Vec<String> {
    int_variable = 15;
    list.push(String::from("addedElement"));
    list = vec!["MutableType_MoveChecker".to_string()];
    // LIST VARIABLE CAN NOT BE USED FOR PASS BY VALUE AS IT DOES NOT HAVE A BUILT IN COPY TRAIT

    println!(
        "\n-- In pass-move function call --\ntmp_int in main: {}\nlist in main: {}",
        int_variable,
        list.join(" ")
    );

    list
}

fn parameter_pass_borrow(list: &mut Vec<String>, int_variable: &mut i32) {
    *int_variable = 15;
    (*list).push(String::from("addedElement"));

    *list = vec!["MutableType_BorrowChecker".to_string()];

    println!(
        "\n-- In pass-borrow function call --\ntmp_int in main: {}\nlist in main: {}",
        int_variable,
        list.join(" ")
    );
}

fn subprogramPass(function: fn(i32), int_variable: i32) {
    // let refEnvChecker: &str = "Ad Hoc Binding"; SYNTAX ERROR WHEN GLOBAL REFENVCHECKER IS COMMENTED OUT
    function(int_variable);
}

fn passedFunction(int_variable: i32) {
    println!("Passed function with a parameter: {}", int_variable);
    println!("Binding refEnvChecker: {}", refEnvChecker);
}